### ISSUE TITLE
Enable non-isothermal simulations with PorousFlowFluidStateBrineCO2

### ISSUE THREAD

**Issue by WilkAndy** [idaholab/moose#10167]
## Rationale

PorousFlowFluidStateBrineCO2 doesn't compute enthalpy or internal energy of the fluids, but it also doesn't compute the porepressure and saturation **real** material properties (only the std::vector versions) so that PorousFlowSingleComponentFluid can't compute the enthalpy or internal energy either.  Hence non-isothermal simulations aren't possible with PorousFlowFluidStateBrineCO2

## Description

We can either compute the real versions of the saturation and porepressure, or we can have a compute_enthalpy, etc, flag to PorousFlowFluidStateBrineCO2.  Or perhaps there's another design.  I'm not sure what the best version is.  What do you think, @cpgr ?

## Impact

Enabling non-isothermal simulations that involve dissolution of gases in liquids.

**Comment by cpgr** [comment]
Oh yeah, I haven't got to that yet. Sorry Andy, I forgot that it hasn't been implemented.

**Comment by WilkAndy** [comment]
@cpgr is going to do this, :-)

**Comment by WilkAndy** [comment]
Prompting @cpgr on this.  It's going to be needed by the ETH geothermal group as they're exploring CO2 + water in geothermal reservoirs.

**Comment by cpgr** [comment]
I've worked out how to do this, and it is on my to do list. Hopefully this month! We would find this useful as well.

**Comment by WilkAndy** [comment]
Cool beans, thanks @cpgr

**Comment by cpgr** [comment]
Isn't it 5am?

**Comment by WilkAndy** [comment]
i keep waking up too early.  At least 5am isn't super unusual for people who like to get to work super early.  3am is just crazy :-(

**Comment by WilkAndy** [comment]
Hey, it'd be really cool if the salt could be a PorousFlow mass fraction.  Then it could advect with the fluid properly.  And why can't we also have more mass fractions hanging around - chemical reactants, tracers, etc, that don't affect the brineco2 physical properties but allow us to do reactions and advection?

**Comment by cpgr** [comment]
OK, now your earlier question makes sense to me! I don't think salt will advect with the CO2-brine class yet. It needs to be included as a variable, but I think that I only added it as a constant to make the initial implementation easier. This is another thing on my todo list - I think I'll try this (and the non-isothermal part) next week as I don't have anything that desperately needs to be finished.

I guess we can add other things like tracers with some other class that just sits alongside this one.

**Comment by WilkAndy** [comment]
Cool, i'm looking forward to seeing this.  I didn't appreciate how broadly useful this BrineCO2 class would be - lots of people in Europe want to use it.

I created a PorousFlowBasicAdvection (PR #11067) that will advect the salt "outside" of the main PorousFlow architecture (using the Darcy velocity of a phase).  For many simulations it'd be sufficient, but it's not that great as there's no upwinding or mass lumping.  I created it really as a framework in which to implement RDG, but for the most part it'd be nicer to have the salt+etc all handled within the PorousFlow framework just like any other mass fraction.
